pkg/querier/queryrange/cache.go

```diff
@@ -0,0 +1,346 @@
+package queryrange
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sort"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk/cache"
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"github.com/opentracing/opentracing-go"
+	otlog "github.com/opentracing/opentracing-go/log"
+	"github.com/prometheus/common/model"
+	"github.com/uber/jaeger-client-go"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/user"
+)
+
+type lokiResultsCache struct {
+	logger   log.Logger
+	cfg      queryrange.ResultsCacheConfig
+	next     queryrange.Handler
+	cache    cache.Cache
+	limits   Limits
+	splitter queryrange.CacheSplitter
+	merger   queryrange.Merger
+}
+
+// NewLokiResultsCacheMiddleware is a cache implementation for "series" and "label" requests
+// this code is mostly taken from cortex's result_cache middleware with only few changes.
+// Main idea to implement custom cache middleware is that with cortex's cache middleware
+// code, we can't correctly cache "series" and "label" responses since they doesn't have a
+// time range associtated with it. This custom code addresses the limitation.
+func NewLokiResultsCacheMiddleware(
```

Comment:
I'm having difficulty finding where this differs from the cortex version -- can you help me out? Also, would it be worth exposing the relevant parts in cortex rather than copy/pasting them here?

Reply:
Most of the logic is just copy/paste. Difference is in `partition` logic. In this implementation of cache, we consider the `extent` only if it within the request range, we don't consider any overlap. For any other request time range which doesn't match this condition we make a  new request.
https://github.com/grafana/loki/blob/0c85a4cdf23976114b786dee045eef5ca2b81de6/pkg/querier/queryrange/cache.go#L187

Whereas cortex just ignores the extents which doesn't overlap
https://github.com/cortexproject/cortex/blob/d16b68152befaabe7c65f501371f20470d0a7bb4/pkg/querier/queryrange/results_cache.go#L397

Agree that we can expose required methods from Cortex